core/aws-core/src/main/java/software/amazon/awssdk/awscore/endpoint/AwsClientEndpointProvider.java

```diff
@@ -156,6 +166,21 @@ private Optional<URI> profileProperty(Builder builder, String profileProperty) {
                                  .flatMap(p -> p.property(profileProperty)));
     }
 
+    private Optional<URI> servicesProperty(Builder builder) {
+        Optional<ProfileFile> profileFile = Optional.ofNullable(builder.profileFile.get());
+        Optional<String> servicesSectionName = profileFile
+            .flatMap(pf -> pf.profile(builder.profileName))
+            .flatMap(p -> p.property("services"));
+
+        Optional<URI> serviceEndpoint = servicesSectionName
+            .flatMap(name -> profileFile.flatMap(pf -> pf.getSection("services", name)))
+            .flatMap(p -> Optional.ofNullable(p.properties().get(builder.serviceProfileProperty
+                                                                 + "." + ProfileProperty.ENDPOINT_URL)))
+            .flatMap(uri -> createUri("services section property", Optional.of(uri)));
```

Comment:
Wrapping uri in an optional is a weird way of doing it, even though it works. I'd have personally just saved the result of the last method as and then returned createUrl(...). Not blocking for you to make a change.